### PR TITLE
Datetimepicker style updates for nimbus

### PR DESCRIPTION
**Pull request by biyeun**
@benrudolph @wpride 

nimbus-datepicker
nimbus-preview-datepicker

### CODE REVIEW

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 1]
```diff
@@ -298,9 +298,10 @@ function DateTimeEntryBase(question, options) {
     maxDate = moment(thisYear + 10, 'YYYY').toDate();
     // Set min date to 100 years in the past
     minDate = moment(thisYear - 100, 'YYYY').toDate();
+    var phoneMode = FormplayerFrontend.request('currentUser').phoneMode;
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'FormplayerFrontend' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 2]
```diff
@@ -314,8 +315,39 @@ function DateTimeEntryBase(question, options) {
                     return;
                 }
                 self.answer(moment(newDate).format(self.serverFormat));
-            }
-        });
+            },
+            onGenerate: function (ct) {
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'ct' is defined but never used (no-unused-vars)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 3]
```diff
@@ -314,8 +315,39 @@ function DateTimeEntryBase(question, options) {
...
+
+                if ($dt.find('.xdsoft_today_button .fa').length < 1) {
+                    $dt.find('.xdsoft_today_button').append($('<i class="fa fa-home" />'));
+                }
+
+                $dt.find('.xdsoft_label i').addClass('fa fa-caret-down');
+
+                if (phoneMode && self.timepicker && self.datepicker) {
+                    $dt.find('.xdsoft_save_selected')
+                        .show().text(django.gettext('Save'))
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'django' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 4]
```diff
@@ -314,8 +315,43 @@ function DateTimeEntryBase(question, options) {
...
+
+                $dt.find('.xdsoft_label i').addClass('fa fa-caret-down');
+
+                if (phoneMode && !self.datepicker && self.timepicker) {
+                    $dt.find('.xdsoft_time_box').addClass('time-box-full');
+                }
+
+                if (phoneMode && self.timepicker && self.datepicker) {
+                    $dt.find('.xdsoft_save_selected')
+                        .show().text(django.gettext('Save'))
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'django' is not defined. (no-undef)

**Comment by biyeun** [comment]
added fullscreen to preview datepicker.
nimbus-datepicker-fullscreen-preview

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 5]
```diff
@@ -314,8 +316,43 @@ function DateTimeEntryBase(question, options) {
...
+
+                $dt.find('.xdsoft_label i').addClass('fa fa-caret-down');
+
+                if (displayOpts.phoneMode && !self.datepicker && self.timepicker) {
+                    $dt.find('.xdsoft_time_box').addClass('time-box-full');
+                }
+
+                if (displayOpts.phoneMode && self.timepicker && self.datepicker) {
+                    $dt.find('.xdsoft_save_selected')
+                        .show().text(django.gettext('Save'))
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'django' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js [hunk 6]
```diff
@@ -404,13 +405,13 @@ WebFormSession.prototype.serverError = function(q, resp) {
...
         self.onload(self, resp);
     });
 };
 
-WebFormSession.prototype.renderFormXml = function (resp, $form) {
+WebFormSession.prototype.renderFormXml = function (resp, $form, displayOptions) {
     var self = this;
     self.session_id = self.session_id || resp.session_id;
-    self.form = Formplayer.Utils.initialRender(resp, self.resourceMap, $form);
+    self.form = Formplayer.Utils.initialRender(resp, self.resourceMap, $form, displayOptions);
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js]: 'Formplayer' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 7]
```diff
@@ -501,3 +538,19 @@ function intpad(x, n) {
...
+/**
+ * Utility that gets the display options from a parent form of a question.
+ * */
+function _getDisplayOptions(question) {
+    var maxIter = 10; // protect against a potential infinite loop
+    var form = question.parent;
+
+    // logic in case the question is in a group or repeat or nested group, etc.
+    while (form.displayOptions === undefined && maxIter > 0) {
+        maxIter --;
```
Comment (benrudolph) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: any chance we can throw an error or warning when this happens. i could see it being a very confusing bug

File: corehq/apps/style/static/cloudcare/less/cloudcare/datepicker.less [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+.xdsoft_datetimepicker .xdsoft_current {
+  background: @cc-brand-mid !important;
+  color: white !important;
+}
+
+.xdsoft_datetimepicker .xdsoft_current.xdsoft_time {
+  box-shadow: darken(@cc-brand-mid, 5) 1px 3px 0 inset !important;
+}
+
+.xdsoft_datetimepicker .xdsoft_mounthpicker .xdsoft_scroller_box {
```
Comment (benrudolph) [inline corehq/apps/style/static/cloudcare/less/cloudcare/datepicker.less]: gotta love typos in libraries

Comment (benrudolph) [inline corehq/apps/style/static/cloudcare/less/cloudcare/datepicker.less]: 😶

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 9]
```diff
@@ -314,8 +316,43 @@ function DateTimeEntryBase(question, options) {
                     return;
                 }
                 self.answer(moment(newDate).format(self.serverFormat));
-            }
-        });
+            },
+            onGenerate: function () {
```
Comment (benrudolph) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: maybe just a sentence on what this function does. not super easy to tell from the code

**Comment by benrudolph** [comment]
this is great! 📅 🕐 ⌛

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 10]
```diff
@@ -313,8 +315,44 @@ function DateTimeEntryBase(question, options) {
...
+
+                $dt.find('.xdsoft_label i').addClass('fa fa-caret-down');
+
+                if (displayOpts.phoneMode && !self.datepicker && self.timepicker) {
+                    $dt.find('.xdsoft_time_box').addClass('time-box-full');
+                }
+
+                if (displayOpts.phoneMode && self.timepicker && self.datepicker) {
+                    $dt.find('.xdsoft_save_selected')
+                        .show().text(django.gettext('Save'))
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'django' is not defined. (no-undef)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js [hunk 11]
```diff
@@ -294,6 +295,7 @@ WebFormSession.prototype.loadForm = function($form, initLang) {
         'nav': 'fao',
         'uses_sql_backend': this.uses_sql_backend,
         'post_url': this.post_url,
+        'oneQuestionPerScreen': true
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js]: Missing trailing comma. (comma-dangle)

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js [hunk 12]
```diff
@@ -313,8 +315,44 @@ function DateTimeEntryBase(question, options) {
...
+
+                $dt.find('.xdsoft_label i').addClass('fa fa-caret-down');
+
+                if (displayOpts.phoneMode() && !self.datepicker && self.timepicker) {
+                    $dt.find('.xdsoft_time_box').addClass('time-box-full');
+                }
+
+                if (displayOpts.phoneMode() && self.timepicker && self.datepicker) {
+                    $dt.find('.xdsoft_save_selected')
+                        .show().text(django.gettext('Save'))
```
Comment (dimagimon) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: 'django' is not defined. (no-undef)

**Comment by biyeun** [comment]
@benrudolph resolved conflicts

**Comment by benrudolph** [comment]
nice, i think we should wait til the dates settle down and merge this after summit. that ok?

**Comment by benrudolph** [comment]
(or at least merge this after today's deploy)

**Comment by biyeun** [comment]
fixed tests

**Comment by biyeun** [comment]
@benrudolph build is ✅

**Comment by biyeun** [comment]
@benrudolph pls merge before one question per screen

File: corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js (hunk 9 above)
Comment (benrudolph) [inline corehq/apps/cloudcare/static/cloudcare/js/form_entry/entrycontrols_full.js]: still would be nice to have a small doc here

**Pull request merged**